Fix: Issues with upgrading colony version

iamsamgibbs opened the pull request:
## How to Test

Requires changes to your dev environment to have a Colony running on version 13. Please follow the steps in #1677 (You will also need to copy over this script into ColonyNetwork as it is missing from the `glwss4` branch: https://github.com/JoinColony/colonyNetwork/blob/develop/scripts/deployOldUpgradeableVersionTruffle.js)


## Description

If already on the latest colony version, the action sidebar will now show an info banner and disable the submit button when the 'upgrade colony version' action type is selected. Metadata has also been adjusted to: `{Upgrade Colony version to v{newVersion} by {initiator}}` and "latest version" in the table, as discussed with Arren.

![Screenshot 2024-01-18 at 13 53 41](https://github.com/JoinColony/colonyCDapp/assets/34915414/d9457e8d-8f24-4c71-9e75-dc36c07ca0a2)

On the advanced settings page, the current colony version is now on the left and latest contract version on the right. Text has been updated from "Last version" to "Latest version".

![Screenshot 2024-01-18 at 13 48 25](https://github.com/JoinColony/colonyCDapp/assets/34915414/8b90f919-b032-4c5f-bba4-2f3145002836)

"Upgrade now" button now opens the action sidebar with the upgrade colony action selected.

![Screenshot 2024-01-18 at 13 48 56](https://github.com/JoinColony/colonyCDapp/assets/34915414/4d868062-d72e-4d6f-b930-69cffcb2db37)

"Upgrade Colony version" link in the admin menu should remain as discussed with Arren.


**Changes** 🏗

* If already on the latest colony version, action sidebar now shows an info banner
* If already on the latest colony version, action sidebar submit button is disabled
* Action sidebar metadata adjusted to `{Upgrade Colony version to v{newVersion} by {initiator}}`
* In the action sidebar form: "New version" updated to "Latest version"
* On the advanced settings page, current colony version and latest contract version now in the correct positions, and "Last version" adjusted to "Latest version".
* "Upgrade now" button on advanced settings page now opens action sidebar

Resolves #1677

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/JoinColony/colonyCDapp?pullRequest=1741) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/JoinColony/colonyCDapp?pullRequest=1741) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/JoinColony/colonyCDapp?pullRequest=1741) it.</sub>

rdig commented:
@area since you're the one that brought this up initially, I'd want you to also test it locally and check if it meets your requirements

rdig commented on src/components/v5/common/ActionSidebar/partials/ActionSidebarContent/partials/SidebarBanner.tsx:
```diff
@@ -88,6 +100,13 @@ export const SidebarBanner: FC = () => {
       ))}
+      {showVersionUpToDateNotification && (
+        <div className="mt-6">
+          <NotificationBanner icon="check-circle" status="success">
```
Why is the status of `NotificationBanner` a plain string and not an `enum`.

_(This is not directed at you @iamsamgibbs, it's just me ranting about the implemetation)_

iamsamgibbs replied:
Opened as new issue: #1770

rdig commented on src/components/v5/common/ActionSidebar/partials/hooks.ts:
```diff
@@ -57,5 +61,9 @@ export const useSubmitButtonDisabled = () => {
     return true;
   }
 
+  if (selectedAction === ACTION.UPGRADE_COLONY_VERSION && !canUpgrade) {
+    return true;
+  }
+
```
This is starting to drag with all the if cases (2 now, more on the way most likely), so please refactor this as a `switch` statement

iamsamgibbs replied:
Refactored this as a switch statement.

rdig reviewed (approved):
Very nice work @iamsamgibbs :100: 

Tested this with actions, working perfectly, however motions were harder to test since I've ran into some issues _(which currently affect `master`)_

I've left a note about refactoring that submit button hook, but otherwise this is good to go.

![Screenshot from 2024-01-19 18-31-44](https://github.com/JoinColony/colonyCDapp/assets/1193222/53b831f6-ba7b-405c-b24d-657152aa4bce)
![Screenshot from 2024-01-19 18-31-57](https://github.com/JoinColony/colonyCDapp/assets/1193222/4107fb71-23e2-420d-972b-97ccd99276d9)

jakubcolony commented on src/components/frame/Extensions/pages/AdvancedPage/AdvancedPage.tsx:
```diff
@@ -48,8 +48,8 @@ const AdvancedPage: FC = () => {
       <ColonyVersionWidget
-        currentVersion={colonyContractVersion}
-        lastVersion={version || 0}
+        currentVersion={version || 0}
```
Might not need the fallback anymore

iamsamgibbs replied:
Removed this unnecessary fallback.

jakubcolony reviewed (approved):
Great work @iamsamgibbs!

rdig commented:
@iamsamgibbs can you please rebase this on the latest `master`. It brings in the motion fixes which'll make this easier to test for @area

area commented:
Creating a motion (after adding the relevant extension) for a colony version upgrade does not appear to work cleanly. The motion is created, and exists, but when trying to see the details of the motion I get the below.

![screenshot-2024-01-24_11-31-03](https://github.com/JoinColony/colonyCDapp/assets/311812/c008c822-7d0c-4c5f-b619-e3ac057135ca)

Having spoken to Raul, he thinks it's likely that there is a switch somewhere in the rendering specifically for motions, and there simply isn't an implementation for a motion that is to upgrade a colony (whereas there is for an action that upgrades a colony, and there are implementations for other motions).

@arrenv, is there a design for this somewhere (i.e. specifically a motion to upgrade a colony)?

rdig reviewed (changes requested):
Checked this with Alex, and indeed, I can reproduce the upgrade via motions issues.

Reverted my approval until that is going to get implemented.

PS: I do realize that this increases the scope of this PR

iamsamgibbs commented:
Added VersionUpgradeMotion as a supported action for the CompletedAction component so the details now show correctly when creating a version upgrade motion.

https://github.com/JoinColony/colonyCDapp/assets/34915414/d2527d7d-a47b-42a1-b72d-4a7ef6b2d7cf

(Unrelated(?) issue: description takes a couple of seconds to pop in - not sure if this happens on other motions?)

---

Also added 'Current version' and 'Latest version' to CompletedAction component which shows when created with permissions or as a motion.

https://github.com/JoinColony/colonyCDapp/assets/34915414/cf1e8d34-405b-4b37-a07a-f20cb5268544

(Unrelated(?) issue: when creating with permissions, title initally appears with the fallback before updating to the correct value. Description takes a couple of seconds to pop in.)

area commented:
Functionality all appears there, both for motions and for doing the upgrade via permissions. For the component, I would not have 'Latest Version', which is only accurate if the colony in question is one version behind, and would be arguably misleading in the future. Perhaps 'Upgrade to Version'?

I did encounter some presumably undesirable behaviour, but probably worthy of its own issue and PR, rather than holding up this one, as it affects all motions:

* Create two motions to upgrade the colony, and move them both to the 'execution' step.
* The first one can be finalized
* The second one cannot (which is the correct contract behaviour), but clicking the 'Finalize' button appears to do nothing in the UI, and logs the `voting-execution-failed-not-one-week` error that occurs when (presumably) estimating gas to the console. Repeatedly clicking the button is possible, and has the same effect each time. I don't _believe_ this is actually broadcasting a transaction though, so users will not be losing gas fees, just getting confused.

iamsamgibbs commented:
Thanks for testing this! I have updated "Latest version" to "New version" as per the figma design.

![Screenshot 2024-01-26 at 12 29 56](https://github.com/JoinColony/colonyCDapp/assets/34915414/f9b0c5cf-588d-4c94-bd83-e5d9f4c2f312)

I was able to recreate the finalising issue and agree it should be raised as a new issue.

area reviewed (dismissed):
:+1:

rdig commented on src/components/v5/shared/ColonyVersionWidget/ColonyVersionWidget.tsx:
```diff
@@ -14,12 +17,24 @@ const displayName = 'v5.ColonyVersionWidget';
+    actionSidebarToggle: [, { toggleOn: toggleActionSidebarOn }],
+  } = useActionSidebarContext();
+
+  const handleUpgradeColony = useCallback(
```
I don't think `useCallback` is really necessary here, as the toggle action does not benefit from being memoized

iamsamgibbs replied:
I've removed this now 👍

rdig reviewed (approved):
Looks good.

Tested both actions and motions and now works as expected. Nice work yet again Sam!

![Screenshot from 2024-01-26 19-23-45](https://github.com/JoinColony/colonyCDapp/assets/1193222/1c1fd702-fd03-4e53-b604-77c3bb988002)
![Screenshot from 2024-01-26 19-25-32](https://github.com/JoinColony/colonyCDapp/assets/1193222/b6e2e27b-a955-4c05-be08-39018d3e3678)
![Screenshot from 2024-01-26 19-25-47](https://github.com/JoinColony/colonyCDapp/assets/1193222/5fffea58-e7f6-4822-986e-3893e198985c)

rdig reviewed (approved):
Re-approved. Nice work.
